File: TermiRustMobile/Models/MobileVaultModels.swift
import Foundation

let mobileVaultSchemaVersion = 1

struct EncryptedMobileVaultEnvelope: Decodable, Equatable {
    let version: Int
    let schemaVersion: Int
    let cipher: String
    let kdf: String
    let salt: String
    let nonce: String
    let ciphertext: String

    enum CodingKeys: String, CodingKey {
        case version
        case schemaVersion = "schema_version"
        case cipher
        case kdf
        case salt
        case nonce
        case ciphertext
    }
}

struct MobileVaultExport: Codable, Hashable {
    let schemaVersion: Int
    let exportId: String
    let createdAtMillis: UInt64
    let updatedAtMillis: UInt64
    let sourceDeviceId: String
    let vaults: [MobileVault]
    let hosts: [MobileHost]
    let groups: [MobileGroup]
    let tags: [String]
    let identities: [MobileIdentityMetadata]
    let knownHosts: [MobileKnownHost]
    let sync: MobileSyncMetadata
    let devices: [MobileDeviceRecord]

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case exportId = "export_id"
        case createdAtMillis = "created_at_millis"
        case updatedAtMillis = "updated_at_millis"
        case sourceDeviceId = "source_device_id"
        case vaults
        case hosts
        case groups
        case tags
        case identities
        case knownHosts = "known_hosts"
        case sync
        case devices
    }
}

struct MobileVault: Codable, Hashable, Identifiable {
    let id: String
    let label: String
    let description: String
    let kind: String
}

struct MobileHost: Codable, Hashable, Identifiable {
    let id: String
    let label: String
    let vaultId: String?
    let group: String
    let tags: [String]
    let host: String
    let port: UInt16
    let username: String
    let auth: MobileAuthMetadata
    let jumpHostId: String?
    let startupDirectory: String?
    let startupCommand: String?
    let startInFiles: Bool
    let persistentSession: MobilePersistentSession
    let terminalScrollbackRows: UInt32?
    let colorTag: String?
    let environment: [MobileEnvironmentVariable]
    let knownHostEndpoint: String?

    enum CodingKeys: String, CodingKey {
        case id
        case label
        case vaultId = "vault_id"
        case group
        case tags
        case host
        case port
        case username
        case auth
        case jumpHostId = "jump_host_id"
        case startupDirectory = "startup_directory"
        case startupCommand = "startup_command"
        case startInFiles = "start_in_files"
        case persistentSession = "persistent_session"
        case terminalScrollbackRows = "terminal_scrollback_rows"
        case colorTag = "color_tag"
        case environment
        case knownHostEndpoint = "known_host_endpoint"
    }
}

struct MobileAuthMetadata: Codable, Hashable {
    let kind: MobileAuthKind
    let identityId: String?
    let secretRef: String?

    enum CodingKeys: String, CodingKey {
        case kind
        case identityId = "identity_id"
        case secretRef = "secret_ref"
    }
}

enum MobileAuthKind: String, Codable, Hashable {
    case password
    case privateKey = "private_key"
}

struct MobilePersistentSession: Codable, Hashable {
    let enabled: Bool
    let sessionName: String?
    let detachOthers: Bool

    enum CodingKeys: String, CodingKey {
        case enabled
        case sessionName = "session_name"
        case detachOthers = "detach_others"
    }
}

struct MobileEnvironmentVariable: Codable, Hashable {
    let name: String
    let value: String
}

struct MobileIdentityMetadata: Codable, Hashable, Identifiable {
    let id: String
    let label: String
    let vaultId: String?
    let kind: String
    let publicKey: String?
    let fingerprint: String?
    let secretRef: String?

    enum CodingKeys: String, CodingKey {
        case id
        case label
        case vaultId = "vault_id"
        case kind
        case publicKey = "public_key"
        case fingerprint
        case secretRef = "secret_ref"
    }
}

struct MobileKnownHost: Codable, Hashable, Identifiable {
    var id: String { endpoint }

    let endpoint: String
    let publicKey: String
    let algorithm: String?
    let fingerprint: String?

    enum CodingKeys: String, CodingKey {
        case endpoint
        case publicKey = "public_key"
        case algorithm
        case fingerprint
    }
}

struct MobileGroup: Codable, Hashable, Identifiable {
    let id: String
    let name: String
}

struct MobileSyncMetadata: Codable, Hashable {
    let revision: UInt64?
    let lastSyncedAtMillis: UInt64?

    enum CodingKeys: String, CodingKey {
        case revision
        case lastSyncedAtMillis = "last_synced_at_millis"
    }
}

struct MobileDeviceRecord: Codable, Hashable, Identifiable {
    var id: String { deviceId }

    let deviceId: String
    let label: String
    let platform: String?
    let publicKey: String?
    let revokedAtMillis: UInt64?

    enum CodingKeys: String, CodingKey {
        case deviceId = "device_id"
        case label
        case platform
        case publicKey = "public_key"
        case revokedAtMillis = "revoked_at_millis"
    }
}

File: app/src/main/java/com/termirust/mobile/data/MobileVaultModels.kt
package com.termirust.mobile.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

const val MOBILE_VAULT_SCHEMA_VERSION = 1

@Serializable
data class EncryptedMobileVaultEnvelope(
    val version: Int,
    @SerialName("schema_version") val schemaVersion: Int,
    val cipher: String,
    val kdf: String,
    val salt: String,
    val nonce: String,
    val ciphertext: String,
)

@Serializable
data class MobileVaultExport(
    @SerialName("schema_version") val schemaVersion: Int,
    @SerialName("export_id") val exportId: String,
    @SerialName("created_at_millis") val createdAtMillis: ULong,
    @SerialName("updated_at_millis") val updatedAtMillis: ULong,
    @SerialName("source_device_id") val sourceDeviceId: String,
    val vaults: List<MobileVault> = emptyList(),
    val hosts: List<MobileHost> = emptyList(),
    val groups: List<MobileGroup> = emptyList(),
    val tags: List<String> = emptyList(),
    val identities: List<MobileIdentityMetadata> = emptyList(),
    @SerialName("known_hosts") val knownHosts: List<MobileKnownHost> = emptyList(),
    val sync: MobileSyncMetadata = MobileSyncMetadata(),
    val devices: List<MobileDeviceRecord> = emptyList(),
    @SerialName("device_keys") val deviceKeys: List<MobileDeviceVaultKey> = emptyList(),
) {
    fun sourceDeviceRecord(): MobileDeviceRecord? =
        devices.firstOrNull { it.deviceId == sourceDeviceId }

    fun sourceDeviceRevoked(): Boolean =
        sourceDeviceRecord()?.revokedAtMillis != null

    fun isOlderThan(other: MobileVaultExport): Boolean {
        val revision = sync.revision
        val otherRevision = other.sync.revision
        if (revision != null && otherRevision != null && revision != otherRevision) {
            return revision < otherRevision
        }
        return updatedAtMillis < other.updatedAtMillis
    }

    fun isDeviceRevoked(deviceId: String): Boolean {
        val trimmed = deviceId.trim()
        if (trimmed.isEmpty()) {
            return false
        }
        return devices.any { it.deviceId == trimmed && it.revokedAtMillis != null }
    }

    fun activeDeviceKey(deviceId: String): MobileDeviceVaultKey? {
        if (isDeviceRevoked(deviceId)) {
            return null
        }
        return deviceKeys.firstOrNull { it.deviceId == deviceId && it.revokedAtMillis == null }
    }
}

@Serializable
data class MobileVault(
    val id: String,
    val label: String,
    val description: String = "",
    val kind: String = "",
)

@Serializable
data class MobileHost(
    val id: String,
    val label: String,
    @SerialName("vault_id") val vaultId: String? = null,
    val group: String = "",
    val tags: List<String> = emptyList(),
    val host: String,
    val port: Int,
    val username: String,
    val auth: MobileAuthMetadata,
    @SerialName("jump_host_id") val jumpHostId: String? = null,
    @SerialName("startup_directory") val startupDirectory: String? = null,
    @SerialName("startup_command") val startupCommand: String? = null,
    @SerialName("start_in_files") val startInFiles: Boolean = false,
    @SerialName("persistent_session") val persistentSession: MobilePersistentSession = MobilePersistentSession(),
    @SerialName("terminal_scrollback_rows") val terminalScrollbackRows: UInt? = null,
    @SerialName("color_tag") val colorTag: String? = null,
    val environment: List<MobileEnvironmentVariable> = emptyList(),
    @SerialName("known_host_endpoint") val knownHostEndpoint: String? = null,
)

@Serializable
data class MobileAuthMetadata(
    val kind: MobileAuthKind,
    @SerialName("identity_id") val identityId: String? = null,
    @SerialName("secret_ref") val secretRef: String? = null,
)

@Serializable
enum class MobileAuthKind {
    @SerialName("password")
    Password,

    @SerialName("private_key")
    PrivateKey,
}

@Serializable
data class MobilePersistentSession(
    val enabled: Boolean = false,
    @SerialName("session_name") val sessionName: String? = null,
    @SerialName("detach_others") val detachOthers: Boolean = false,
)

@Serializable
data class MobileEnvironmentVariable(
    val name: String,
    val value: String,
)

@Serializable
data class MobileIdentityMetadata(
    val id: String,
    val label: String,
    @SerialName("vault_id") val vaultId: String? = null,
    val kind: String,
    @SerialName("public_key") val publicKey: String? = null,
    val fingerprint: String? = null,
    @SerialName("secret_ref") val secretRef: String? = null,
)

@Serializable
data class MobileKnownHost(
    val endpoint: String,
    @SerialName("public_key") val publicKey: String,
    val algorithm: String? = null,
    val fingerprint: String? = null,
)

@Serializable
data class MobileGroup(
    val id: String,
    val name: String,
)

@Serializable
data class MobileSyncMetadata(
    val revision: ULong? = null,
    @SerialName("last_synced_at_millis") val lastSyncedAtMillis: ULong? = null,
)

@Serializable
data class MobileDeviceRecord(
    @SerialName("device_id") val deviceId: String,
    val label: String,
    val platform: String? = null,
    @SerialName("public_key") val publicKey: String? = null,
    @SerialName("paired_at_millis") val pairedAtMillis: ULong? = null,
    @SerialName("last_seen_at_millis") val lastSeenAtMillis: ULong? = null,
    @SerialName("revoked_at_millis") val revokedAtMillis: ULong? = null,
)

@Serializable
data class MobileDeviceVaultKey(
    @SerialName("key_id") val keyId: String,
    @SerialName("device_id") val deviceId: String,
    @SerialName("wrapping_algorithm") val wrappingAlgorithm: String,
    @SerialName("encrypted_vault_key") val encryptedVaultKey: String,
    @SerialName("created_at_millis") val createdAtMillis: ULong? = null,
    @SerialName("revoked_at_millis") val revokedAtMillis: ULong? = null,
)
